Overhaul Jib AST

janvrany opened the pull request:
This PR is an attempt to overhaul Jib AST and decouple it from code supporting interpretation and/or reasoning.

janvrany committed:
Fix parsing hexadecimal bit literals

This commit fixes a bug in parsing bit literals (`Exp_Bits`). Hex bits
literals are treated just like binary bits literals - produce BV of
fixed bit-length (based on length of literal's string representation,
this means that leading zeros are significant).

This renders class `AnyBitVector` unnecessary so it is deleted. While
at it, this commit also refactors the code not to rely on extension
`BitVector class >> #parser0b` - the less extensions the better. Finally
it adds `Exp_Bits >> #printOn:` for easier debugging.

janvrany committed:
Remove unnecessary class `Box`

janvrany committed:
Introduce new abstract base class `JibProgramNode`

This commit makes allk Jib parse nodes to inherit from (new) base class
`JibProgramNode`. This is a first step in adding "standard AST stuff"
like parent-child relationship, pointers back to Jib source code and so
on.

janvrany committed:
Add #isAbstract marker to abstract node classes

This commit marks all abstract base classes for Jib AST nodes with
`#isAbstract` for use by tooling.

janvrany committed:
Do not convert Jib types to Sprite types while parsing

This commit stops converting "Jib" types to "Sprite" types when parsing.
This conversion can be done any time later when needed.

Therefore all nodes that contain type information now hold on `Ty`
sub-instance rather than on `RType`.

shingarov commented on src/Sail-Jib-Parsing/PPParser.extension.st:
Can we somehow stop hating the hole so much?

shingarov reviewed (changes requested):
LGTM modulo the place where the hole-killing is sneaking into an otherwise-unrelated PR.

janvrany committed:
Rename instvars in Jib AST nodes to match ir_parser.lalrpop

This commit rename instance variables in `Def_*` nodes to match those
used in ir_parser.lalrpop around line 197 (commit 1ac01cd [1]).

I personally think this is a bad idea. These are not the names of members
of individual "structs", they are "local" variable names in parser
production action code. Worse, they're named after production rule name,
not the meaning.

I bet one would read code accessing 'ids' and wondering what is "ids"
only to find out that in `Def_Fn` its the argument names whereas in
`Def_Struct` is is something else (and not even a list of ids!) hundred
times more often than grepping `ir_parser.lalrpop`. Especially because
there's no comment whatsoever to suggest that.

[1] https://github.com/rems-project/isla/blob/1ac01cd5ecb111630de744e651d9fee5a8127af9/isla-lib/src/ir_parser.lalrpop

janvrany committed:
Remove use of reader when parsing

So far reader was used to deal with parsing (mutually) recursive
structures. This commit deals with the problem differently:

In addition to #inducedParser class method that defines how to parse
the structure it it introduces new (class) method `#asParser` following
the PPParser which is the to use in other `#inducedParser` methods to
parse constituents.

The key trick is that `#asParser` temporarily creates a delegate parser
(if not already) and then sets its contents to result if #inducedParser.
If this method is called recursively again (as part of #inducedParser),
then the delegate is already there and it is simply returned.

The advantage of this is that it is more readable and does not require
doing anything special if the structure may be recursive (which may not
be obvious).

While at it, make `JibProgramNode class >> #inducedParser` abstract and
get rid of `#adtSummands` - this just makes the code even simpler while
not significantly more work (only one or two new #inducedParser` were
added).

janvrany committed:
Remove use of `JibFnBody`

This commit refactors `Def_Fn` not to use `JibFnBody` and hold on
instructions directly. While at it, remove `body` instvar to `instrs`
(which is in line with upstream grammar/parser).

The class `JibFnBody` itself is kept around and will be deleted in a
commit with all other old stuff we decide to delete and build anew.

janvrany committed:
Remove old code from Jib AST nodes

This commit removes all "old" code supporting both interpretation and
verification from AST node classes. This code need to be re-grown back,
this time properly engineered and using variable bit vector support in
MA.

The code for creating CFG of a Jib function and for computing liveness
is kept. All other non-AST classes and code there is also kept.

janvrany committed:
Refactor JibProgram

This commit reimplements `JibProgram` in similar fashion as all the other
nodes: it define its own #inducedParser` and all definitions are chucked
into a single list of definitions.

This renders `JibParser` more or less unnecessary.

janvrany committed:
Move the meat of `#constructNewFrom:accordingTo:` to instance side

janvrany committed:
Refactor JibSourceLoc

This commit refactors `JibSourceLoc`, making it abstract and introducing
3 subclasses, `SourceLoc_Known`, `SourceLoc_Unknown` and
`SourceLoc_UnknownUnique`. This matches upstream isla-lib [1], modulo
naming (since upstream itself is bit inconsistent in naming and
code structure).

[1]: https://github.com/rems-project/isla/blob/447e0de6e48972f40e72cd1005183ced96da0b66/isla-lib/src/ir_parser.lalrpop#L163-L175

janvrany committed:
Refactor "singleton nodes"

This commit refactor a number of "singleton" nodes and makes them
non-singletons. This is a preparation for introducing parent/chilren
relationship. Also we may want to record reference to *Jib* source for
debugging and error handling and to do that node instances cannot be
shared.

janvrany committed:
Refactor type nodes

This commit fixes a number of `Ty` subclasses to parse properly.

janvrany committed:
Fix parsing of `#mangle` and `#tuplestruct` pragmas

This commit fixes parsing of pragmas: they're prefixed with `#`.

janvrany committed:
Add parent/chilren relationship among nodes

This commit establishes parent/children relationship among nodes. This
enables AST traversal.

janvrany committed:
Fix parsing of union and struct definitions

janvrany committed:
Fix parsing of `Instr_Init`

janvrany committed:
Fix parsing of `Exp_Field`

janvrany committed:
Fix parsing of `Def_Register`

janvrany committed:
Fix parsing of `Ty_Ref`

janvrany committed:
Fix parsing of `Ty_List`

janvrany committed:
Sidestep left recursion in `Exp_Field`

This commit partially fix parsing of `Exp_Field`. This one is tricky
since it left-recursive. In the wild so far we have seen only two
variants:

   * `id.field`
   * `id as typename.field`

This commit sidestep fixing left recursion by only caring for these two
cases. May work until it doesn't.

janvrany committed:
Fix parsing of `Exp_Struct`

shingarov commented on src/Sail-Jib/Exp_Field.class.st:
```diff
@@ -10,7 +10,7 @@ Class {
 { #category : #syntax }
 Exp_Field class >> inducedParser [
-	^ Exp_Id asParser "this should be Exp; however, this leads to recursion on the left"
+	^ (Exp_Unwrap asParser / Exp_Id asParser) "this should be Exp; however, this leads to recursion on the left"
```
Can we put a "BOGUS" marker inside that comment?

shingarov reviewed (changes requested):
LGTM.  I still like to mark sidestepped problems explicitly: so that when it comes back and bites me, I look around that area of the code and immediately remember, "aha!  I remember I postponed this for later, now is the time there is no wiggling out of implementing it properly".

janvrany committed:
Add "BOGUS" marker to `Exp_Field`, `Exp_Kind` and `Exp_Unwrap`

...as they parse only a subset of what Jib allows to express.
